src/components/modal/index.js

```diff
@@ -0,0 +1,80 @@
+import BaseComponent from "../base-component.js";
+
+import "./modal-style.css";
+
+export default class Modal extends BaseComponent {
+  status = "closed";
+
+  onDocumentClick = (event) => {
+    const isModal = event.target.closest('[data-element="modalContainer"]');
+
+    if (!isModal) {
+      this.close();
+    }
+  };
+
+  onDocumentKeyDown = (event) => {
+    if (event.code === "Escape") {
+      event.preventDefault();
+      this.close();
+    }
+  };
+
+  constructor(component = {}) {
+    super();
+    this.component = component;
+    this.controlSelector = "hidden";
+    this.init();
+    this.renderComponent();
+  }
+
+  addComponent(component = {}) {
+    this.component = component;
+    this.renderComponent();
+  }
+
+  get template() {
+    return `<div class="app-modal hidden" data-cy="app-modal">
+      <div class="modal__content" data-element="modalContainer">
+        <div data-element="content"></div>
+      </div>
+    </div>`;
+  }
+
+  renderComponent() {
+    const { content } = this.subElements;
+
+    if (Object.keys(this.component).length) {
+      content.append(this.component.element);
+    }
+  }
+
+  open() {
+    document.body.append(this.element);
+
+    this.addEventListeners();
+    this.element.classList.remove(this.controlSelector);
```

Comment:
**Performance Issue**: The 'open' method directly manipulates the DOM and modifies class attributes, which can cause reflows and repaints, impacting performance when opening the modal. <br> **Fix**: Optimize the open method to reduce layout thrashing by batching DOM read/write operations and using requestAnimationFrame for non-urgent updates. <br> **Code Suggestion**: 
 ```
 +    open() {
 +      requestAnimationFrame(() => {
 +        document.body.append(this.element);
 +        this.addEventListeners();
 +        this.element.classList.remove(this.controlSelector);
 +      });
 +    }
 ```